types/nearest-color/index.d.ts

```diff
@@ -0,0 +1,173 @@
+ * nearestColor('#f88');                   // => '#f80'
+ * nearestColor('#ffe');                   // => '#ff0'
+ * nearestColor('#efe');                   // => '#ff0'
+ * nearestColor('#abc');                   // => '#808'
+ * nearestColor('red');                    // => '#f00'
+ * nearestColor('foo');                    // => throws
+ * ```
+ */
+declare const nearestColor: NearestColor<ColorMatch | string>;
+
+interface NearestColor<TReturnType extends ColorMatch | string> {
+    (needle: NeedleColor, colors?: readonly ColorSpec[]): TReturnType | null;
+}
+
+/**
+ * Provides a matcher to find the nearest color based on the provided list of
+ * available colors.
+ *
+ * @param availableColors An array of hex-based color
+ *     strings, or an object mapping color *names* to hex values.
+ * @return A function with the same behavior as {@link nearestColor}, but with the list of colors predefined.
+ *
+ * @example
+ * ```js
+ * const colors = {
+ *   'maroon': '#800',
+ *   'light yellow': { r: 255, g: 255, b: 51 },
+ *   'pale blue': '#def',
+ *   'white': 'fff'
+ * };
+ *
+ * const bgColors = [
+ *   '#eee',
+ *   '#444'
+ * ];
+ *
+ * const invalidColors = {
+ *   'invalid': 'foo'
+ * };
+ *
+ * const getColor = nearestColor.from(colors);
+ * const getBGColor = getColor.from(bgColors);
+ * const getAnyColor = nearestColor.from(colors).or(bgColors);
+ *
+ * getColor('ffe');
+ * // => { name: 'white', value: 'fff', rgb: { r: 255, g: 255, b: 255 }, distance: 17}
+ *
+ * getColor('#f00');
+ * // => { name: 'maroon', value: '#800', rgb: { r: 136, g: 0, b: 0 }, distance: 119}
+ *
+ * getColor('#ff0');
+ * // => { name: 'light yellow', value: '#ffff33', rgb: { r: 255, g: 255, b: 51 }, distance: 51}
+ *
+ * getBGColor('#fff'); // => '#eee'
+ * getBGColor('#000'); // => '#444'
+ *
+ * getAnyColor('#f00');
+ * // => { name: 'maroon', value: '#800', rgb: { r: 136, g: 0, b: 0 }, distance: 119}
+ *
+ * getAnyColor('#888'); // => '#444'
+ *
+ * nearestColor.from(invalidColors); // => throws
+ * ```
+ */
+declare const from: NearestColorFrom;
+
+interface NearestColorFrom {
+    (availableColors: readonly string[]): NearestColor<string>;
+    (availableColors: { [colorName: string]: string }): NearestColor<ColorMatch>;
+}
+
+/**
+ * A map from the names of standard CSS colors to their hex values.
+ */
+declare const STANDARD_COLORS: { [colorName: string]: string };
+
+/**
+ * Default colors. Comprises the colors of the rainbow (good ol' ROY G. BIV).
+ * This list will be used for calls to {@nearestColor} that don't specify a list
+ * of available colors to match.
+ */
+declare const DEFAULT_COLORS: readonly ColorSpec[];
+
+declare const nearestColorNamespace: {
+    nearestColor: typeof nearestColor;
+    from: typeof from;
+    STANDARD_COLORS: typeof STANDARD_COLORS;
+    DEFAULT_COLORS: typeof DEFAULT_COLORS;
+    VERSION: string;
+};
+
+export = nearestColorNamespace;
+
+export as namespace nearestColorNamespace;
```

Comment:
```suggestion
export as namespace nearestColor;
```
this seems to be a namespace/global symbol name. If I'm correct, please amend other places.
e.g. check this:
https://github.com/dtao/nearest-color/blob/master/nearestColor.js#L400
(where context seems to be 'window' or 'global', etc IMO

Reply:
WDYM besides the namespace change? AFAIU this is only applicable when the package is consumed as a UMD (or similar) script, so I think your suggestion would be all that needs to be changed.
Am I missing something?

Reply:
`export as namespace` creates global variable, so I meant something like:
```ts
declare namespace _ {
    interface RGB {
        /**
         * The red component, from 0 to 255
         */
        r: number;
        /**
         * The green component, from 0 to 255
         */
        g: number;
        /**
         * The blue component, from 0 to 255
         */
        b: number;
    }

    interface ColorSpec {
        /**
         * A name for the color, e.g., 'red'
         */
        name?: string;
        /**
         * The hex-based color string, e.g., '#FF0'
         */
        source: string;
        /**
         * The {@link RGB} color values
         */
        rgb: RGB;
    }

    interface ColorMatch {
        name: string;
        value: string;
        rgb: RGB;
        distance: number;
    }

    type NeedleColor = RGB | string;

    interface NearestColor<TReturnType extends ColorMatch | string> {
        (needle: NeedleColor, colors?: readonly ColorSpec[]): TReturnType | null;
    }

    const from: NearestColorFrom;

    interface NearestColorFrom {
        (availableColors: readonly string[]): NearestColor<string>;
        (availableColors: { [colorName: string]: string }): NearestColor<ColorMatch>;
    }

    const STANDARD_COLORS: { [colorName: string]: string };

    const DEFAULT_COLORS: readonly ColorSpec[];
}

declare const nearestColor: _.NearestColor<_.ColorMatch | string> & typeof _;

export = nearestColor;

export as namespace nearestColor;
```
so it can be used both CJS and as global script:
```ts
import nearestColor = require("nearest-color");
import { DEFAULT_COLORS, from } from "nearest-color";

nearestColor("#800");

from(["#800"])("#fff");
```
```ts
nearestColor("#800");
nearestColor.from(["#800"])("#fff");
nearestColor.from(["#800"])("#fff", nearestColor.DEFAULT_COLORS);
...
```

Reply:
Thanks, this makes a lot of sense. I updated the code in the latest revision.